Using the value O.this where O is an object does not work

```scala
object O {
   val x: this.type = O.this
}
```
This compiles with scalac but fails with dotty:
```
Exception in thread "main" java.lang.AssertionError: assertion failed: found: <noprefix>, expected: O$.type(O)
at scala.Predef$.assert(Predef.scala:165)
at dotty.tools.dotc.typer.Implicits$ImplicitSearch.<init>(Implicits.scala:455)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1$$anonfun$apply$17.apply(Implicits.scala:432)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1$$anonfun$apply$17.apply(Implicits.scala:428)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1.apply(Implicits.scala:428)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1.apply(Implicits.scala:424)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Implicits$class.inferImplicit(Implicits.scala:424)
at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:58)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferView$1.apply(Implicits.scala:407)
at dotty.tools.dotc.typer.Implicits$$anonfun$inferView$1.apply(Implicits.scala:401)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Implicits$class.inferView(Implicits.scala:400)
at dotty.tools.dotc.typer.Typer.inferView(Typer.scala:58)
at dotty.tools.dotc.typer.Typer.adaptToSubType$1(Typer.scala:1339)
at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:1311)
at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:1374)
at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1$$anonfun$apply$14.apply(Typer.scala:1163)
at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1$$anonfun$apply$14.apply(Typer.scala:1160)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1.apply(Typer.scala:1160)
at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1.apply(Typer.scala:1160)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1159)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1068)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1066)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1066)
at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1105)
at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:860)
at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:854)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:854)
at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1003)
at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1056)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1068)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1066)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1066)
at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1090)
at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1101)
at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:902)
at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:884)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:884)
at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1009)
at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1056)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1068)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1066)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1066)
at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1090)
at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1101)
at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:949)
at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:940)
at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:940)
at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1046)
at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1058)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1068)
at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1066)
at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:143)
at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:51)
at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1066)
at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1105)
at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:41)
at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:18)
at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:39)
at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:52)
at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:52)
at scala.collection.immutable.List.foreach(List.scala:381)
at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:52)
at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:49)
at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:47)
at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply$mcV$sp(Run.scala:47)
at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:39)
at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:39)
at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
at dotty.tools.dotc.Run.compileSources(Run.scala:38)
at dotty.tools.dotc.Run.compile(Run.scala:29)
at dotty.tools.dotc.Driver.doCompile(Driver.scala:20)
at dotty.tools.dotc.Main$.doCompile(Main.scala:26)
at dotty.tools.dotc.Driver.process(Driver.scala:31)
at dotty.tools.dotc.Driver.main(Driver.scala:40)
at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/src/dotty/tools/dotc/typer/Implicits.scala
+++ b/src/dotty/tools/dotc/typer/Implicits.scala
@@ -402,7 +402,8 @@ trait Implicits { self: Typer =>
         || (to isRef defn.ObjectClass)
         || (to isRef defn.UnitClass)
         || (from.tpe isRef defn.NothingClass)
-        || (from.tpe isRef defn.NullClass)) NoImplicitMatches
+        || (from.tpe isRef defn.NullClass)
+        || (from.tpe eq NoPrefix)) NoImplicitMatches
     else
       try inferImplicit(to.stripTypeVar.widenExpr, from, from.pos)
       catch {

--- a/src/dotty/tools/dotc/typer/TypeAssigner.scala
+++ b/src/dotty/tools/dotc/typer/TypeAssigner.scala
@@ -8,6 +8,7 @@ import Scopes._, Contexts._, Constants._, Types._, Symbols._, Names._, Flags._, 
 import ErrorReporting._, Annotations._, Denotations._, SymDenotations._, StdNames._, TypeErasure._
 import util.Positions._
 import config.Printers._
+import NameOps._
 
 trait TypeAssigner {
   import tpd._
@@ -16,7 +17,11 @@ trait TypeAssigner {
    *  @param packageOk   The qualifier may refer to a package.
    */
   def qualifyingClass(tree: untpd.Tree, qual: Name, packageOK: Boolean)(implicit ctx: Context): Symbol = {
-    def qualifies(sym: Symbol) = sym.isClass && (qual.isEmpty || sym.name == qual)
+    def qualifies(sym: Symbol) =
+      sym.isClass && (
+          qual.isEmpty ||
+          sym.name == qual ||
+          sym.is(Module) && sym.name.stripModuleClassSuffix == qual)
     ctx.outersIterator.map(_.owner).find(qualifies) match {
       case Some(c) if packageOK || !(c is Package) =>
         c

--- a/test/dotc/tests.scala
+++ b/test/dotc/tests.scala
@@ -155,6 +155,7 @@ class tests extends CompilerTest {
   @Test def neg_i0091_infpaths = compileFile(negDir, "i0091-infpaths", xerrors = 3)
   @Test def neg_i0248_inherit_refined = compileFile(negDir, "i0248-inherit-refined", xerrors = 4)
   @Test def neg_i0281 = compileFile(negDir, "i0281-null-primitive-conforms", xerrors = 3)
+  @Test def neg_i324 = compileFile(negDir, "i324", xerrors = 2)
   @Test def neg_i583 = compileFile(negDir, "i0583-skolemize", xerrors = 2)
   @Test def neg_i941 = compileFile(negDir, "i941", xerrors = 3)
   @Test def neg_finalSealed = compileFile(negDir, "final-sealed", xerrors = 2)

--- a/tests/neg/i324.scala
+++ b/tests/neg/i324.scala
@@ -1,0 +1,5 @@
+class O
+object O {
+   val x: this.type = OO.this
+   val y: O = OO.this
+}

--- a/tests/pos/i324.scala
+++ b/tests/pos/i324.scala
@@ -1,0 +1,7 @@
+class O
+object O {
+   val x: this.type = O.this
+}
+object O2 {
+   val x: this.type = O2.this
+}

